### PR TITLE
Initial experimental support for deferred actions when count or for_each are unknown

### PR DESCRIPTION
**Pull request by apparentlymart**
This PR represents the first real functional milestone toward solving https://github.com/hashicorp/terraform/issues/30937, in a very early experimental capacity.

There are still various details not quite right here yet, but this pile of commits has already got big enough so I think it's time to land it and then we can continue working on the details in subsequent smaller PRs.

The end result of all of this is that:
- In any module that's opted in to the `unknown_instances` language experiment, it's valid for `count` or `for_each` in a `module` or `resource` block to be unknown. (That was actually mostly true before, but there was a bug making it not work for module calls, and turning it on before just caused the plan to be broken rather than doing anything useful.)
- For any `resource` block that has an unknown `count` or `for_each`, or that is declared beneath a module call with unknown `count` or `for_each`, Terraform asks the provider to confirm that what we know of the configuration is plannable, but then just makes a note that all of the instances of this resource must have their actions deferred to a future round.
- For any resource that refers to a resource that's already been deferred for another reason, all of the planning actions run as normal but then the result isn't actually added to the plan, and instead Terraform again just makes a note that it's been deferred to a future round.
- When a module used as part of a component in a stack produces any deferred actions, any downstream components have _all_ of their actions forced deferred, to ensure that we preserve the dependency ordering between components. However, the planning actions do otherwise still run as normal, so that operators can get feedback if we can prove that anything about the configuration is invalid with the information available.

My main priority in this first round was to keep the new codepaths as isolated as possible from the existing code, to minimize the risk of impacting the behavior for anyone not participating in the experiment. Therefore there are some rough edges and compromises called out in comments throughout. Since none of this new code should be reachable in stable releases, I'd ask that reviewers pardon my dust and trust that we'll continue improving this in future PRs.

With that said though: **if you _do_ see something that seems like it could break non-experimental usage, please call that out!** (The fact that I needed to make very few changes to existing tests gives me confidence, but it can't hurt to double-check.)

The _new_ test coverage here is also relatively light, and I apologize for that. My next step here is going to be to try this with a more realistic stack configuration from our set of examples from stacks private preview, and so that will be my primary testing vehicle immediately, and then I'll add more automated tests gradually in future PRs.

I tried to make up for the fact that this is a huge change by breaking it into digestible commits, so I suggest that reviewers take it on a commit-by-commit basis rather than trying to review the entire diff at once.

### CODE REVIEW

File: internal/plans/deferring/deferred.go
Comment (alisdair) [inline internal/plans/deferring/deferred.go:153]: I was a little confused by this paragraph of commentary (and the corresponding one below before the panic). I think the contract we're enforcing here is that callers are required to call these two methods in a specific order for each resource instance, at most once each:

- First call `ShouldDeferResourceInstanceChanges`
- If it returns true, then you must call `ReportResourceInstanceDeferred`

The contract is to crash upon calling these in any other order, or any more often. That's a severe restriction, which may be entirely reasonable, but it was a bit hard for me to follow and I wonder if it's a possible maintenance hazard.

Assuming that I've understood what's happening, which is quite possibly not the case, I have two ideas:

- Bundle these two methods into one, so that it's impossible to call them in the wrong order. Something like:

    ```go
    func (d *Deferred) ShouldDeferResourceInstanceChanges(addr addrs.AbsResourceInstance) bool {
        ret := d.shouldDeferResourceInstanceChanges(addr)
        if ret {
            d.reportResourceInstanceDeferred(addr)
        }
        return ret
    }
    ```

- Or, update the comments throughout to clarify the expectation that these methods are always to be called in this specific order, as a pair, and only once per addr, because we require that the graph walk only evaluates deferrals once per instance.

File: internal/terraform/context_apply2_test.go
Comment (alisdair) [inline internal/terraform/context_apply2_test.go:2812]: This is a neat idea for a test case! I had forgotten about the helpful overlap between refinements and deferred actions so seeing it called out here is great.

**Review by alisdair — APPROVED** [review]
Seems good! None of my comments are of significance and shouldn't block merge. My focus in review was on a possible leak of the experimental behaviour, and I found none.

File: internal/plans/deferring/deferred.go
Comment (apparentlymart) [reply internal/plans/deferring/deferred.go:115]: I did specifically consider adding an "opts" argument to `NewDeferred`, but it seemed weird to complicate that signature for what is currently just a single boolean flag.

I suppose a separate constructor would be a viable approach too, but honestly I decided not to sweat it too much for now. It is true though that this design does make it possible for someone to accidentally misuse this at runtime after initial construction, and so might be worth revisiting this in a later iteration if it seems likely that we're going to continue following this approach once we've got feedback from the experiment.

File: internal/plans/deferring/deferred.go
Comment (apparentlymart) [reply internal/plans/deferring/deferred.go:153]: I agree that this is a little confusing, and I think a big part of what makes it confusing is that many different resource instances are potentially in play, and this constraint only applies for _one specific resource instance address_. In practice then, the sequence of calls is:

1. Some upstream graph node earlier called `ReportResourceInstanceDeferred` to report that _it_ has deferred a change, regardless of why that happened.
2. The current graph node then calls `ShouldDeferResourceInstanceChanges` with its _own_ address, which causes this function to check whether that any of address's upstream dependencies were deferred. (In the case of this example, yes.)
3. The current graph node then responds to that result by calling `ReportResourceInstanceDeferred` with its _own_ resource instance address. (This would then be the step 1 of a downstream incarnation of this sequence, if any)

Bundling the check in with the report is an interesting idea. I think we will still need a separate exported `ReportResourceInstanceDeferred` because in the end this will not be the _only_ reason for a resource instance graph node to need to report that it made a deferral decision: it would do so also if a provider responded to `PlanResourceChange` by saying "I cannot plan this yet", once we get to that functionality in a later PR. But I admittedly cannot think of any reason why one would call `SourceDeferResourceInstanceChanges`, get a `true` response, and _not_ immediately call `ReportResourceInstanceDeferred`, so it does seem plausible to build it so that asking the question and performing the side-effect of reporting are an atomic action from the caller's perspective.

(I do typically try to avoid writing functions that are both accessors and mutators at the same time as far as that's possible to do so, but I do agree that in this case the separation seems more likely to encourage bugs than combining them would.)

File: internal/terraform/context_apply2_test.go
Comment (apparentlymart) [reply internal/terraform/context_apply2_test.go:2812]: Honestly I didn't actually explicitly intend to capture this here... it just emerged naturally from how I constructed the test fixture! But I did quite like to see it once I saw it, so I kept it. :grinning:

File: internal/instances/expander.go
Comment (apparentlymart) [reply internal/instances/expander.go:421]: Thanks! I'll fix this one before merge just because it's not likely to cascade down onto other commits.

(For your other feedback, I'm going to keep it in mind for later work but leave it unchanged for now just to avoid churning this big pile of commits too much, since it's all experimental code we're explicitly intending to revisit anyway.)

Comment (apparentlymart) [reply internal/instances/expander.go:421]: (For consistency I should really have named this method `GetResourceInstanceKeys`, but since changing that would impact the caller from another commit I'm just noting that in a comment here so I can refer back to it while planning later work.)

**Pull request merged**
